### PR TITLE
*: support new 'log' format for trace

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

Give the trace log view to handle the debug or troubleshooting situation that  `format=row` doesn't cover.

through trace log view we can:

-  insight event timeline of the fine-gain event inside one span order by time
-  saw more property(e.g.  request address, region-id .... in tag, log message)

tree in cmd give us the overview of execution but limited by the visual ability couldn't provide execution detail in easy, log format more focus on the event detail and help identify the question.

### What is changed and how it works?

- add `format='log'` to trace command
- add `logutil.Event` and `logutil.SetTag` api
- fix the question that "trace command cannot trace error operation"
- demo in insert/replace

for example trace a replace

```
mysql> trace format='log' replace  into t2(a, b) select a, b from t;
+----------------------------+-----------------------------------------------------------------------------------------------------------------------+----------------------------------+------------------------------------+
| time                       | event                                                                                                                 | tags                             | spanName                           |
+----------------------------+-----------------------------------------------------------------------------------------------------------------------+----------------------------------+------------------------------------+
| 2019-08-08 15:46:03.463555 | --- start span trace ----                                                                                             |                                  | trace                              |
| 2019-08-08 15:46:03.463559 | --- start span session.Execute ----                                                                                   |                                  | session.Execute                    |
| 2019-08-08 15:46:03.463563 | execute: replace  into t2(a, b) select a, b from t                                                                    |                                  | session.Execute                    |
| 2019-08-08 15:46:03.463564 | --- start span session.getTxnFuture ----                                                                              |                                  | session.getTxnFuture               |
| 2019-08-08 15:46:03.463569 | --- start span session.ParseSQL ----                                                                                  |                                  | session.ParseSQL                   |
| 2019-08-08 15:46:03.463589 | --- start span executor.Compile ----                                                                                  |                                  | executor.Compile                   |
| 2019-08-08 15:46:03.463711 | --- start span session.runStmt ----                                                                                   |                                  | session.runStmt                    |
| 2019-08-08 15:46:03.463960 | --- start span executor.handleNoDelayExecutor ----                                                                    |                                  | executor.handleNoDelayExecutor     |
| 2019-08-08 15:46:03.463963 | --- start span *executor.ReplaceExec.Next ----                                                                        |                                  | *executor.ReplaceExec.Next         |
| 2019-08-08 15:46:03.463967 | --- start span *executor.TableReaderExecutor.Next ----                                                                |                                  | *executor.TableReaderExecutor.Next |
| 2019-08-08 15:46:03.463974 | table scan table: t, range: [[-inf,+inf]]                                                                             |                                  | *executor.TableReaderExecutor.Next |
| 2019-08-08 15:46:03.463986 | send Cop request to region 8 at 127.0.0.1:9191                                                                        |                                  | session.runStmt                    |
| 2019-08-08 15:46:03.464253 | receive *tikvpb.BatchCommandsResponse_Response_Coprocessor response with other 1 batched requests from 127.0.0.1:9191 |                                  | session.runStmt                    |
| 2019-08-08 15:46:03.464301 | --- start span *executor.TableReaderExecutor.Next ----                                                                |                                  | *executor.TableReaderExecutor.Next |
| 2019-08-08 15:46:03.464305 | table scan table: t, range: [[-inf,+inf]]                                                                             |                                  | *executor.TableReaderExecutor.Next |
| 2019-08-08 15:46:03.470200 | --- start span session.CommitTxn ----                                                                                 |                                  | session.CommitTxn                  |
| 2019-08-08 15:46:03.470203 | --- start span session.doCommitWitRetry ----                                                                          |                                  | session.doCommitWitRetry           |
| 2019-08-08 15:46:03.470230 | send Prewrite request to region 8 at 127.0.0.1:9191                                                                   | map[commitTs:410320991278596101] | session.doCommitWitRetry           |
| 2019-08-08 15:46:03.470413 | receive *tikvpb.BatchCommandsResponse_Response_Prewrite response with other 1 batched requests from 127.0.0.1:9191    | map[commitTs:410320991278596101] | session.doCommitWitRetry           |
| 2019-08-08 15:46:03.470422 | start get commit ts                                                                                                   | map[commitTs:410320991278596101] | session.doCommitWitRetry           |
| 2019-08-08 15:46:03.470550 | finish get commit ts                                                                                                  | map[commitTs:410320991278596101] | session.doCommitWitRetry           |
| 2019-08-08 15:46:03.470570 | send Commit request to region 8 at 127.0.0.1:9191                                                                     | map[commitTs:410320991278596101] | session.doCommitWitRetry           |
| 2019-08-08 15:46:03.473095 | receive *tikvpb.BatchCommandsResponse_Response_Commit response with other 1 batched requests from 127.0.0.1:9191      | map[commitTs:410320991278596101] | session.doCommitWitRetry           |
| 2019-08-08 15:46:03.473149 | execute done, modify row: 3                                                                                           |                                  | trace                              |
+----------------------------+-----------------------------------------------------------------------------------------------------------------------+----------------------------------+------------------------------------+
24 rows in set (0.01 sec)
```

trace a insert into select:

```
mysql> trace format='log' replace  into t2(a, b) select a, b from t;
+----------------------------+-----------------------------------------------------------------------------------------------------------------------+----------------------------------+------------------------------------+
| time                       | event                                                                                                                 | tags                             | spanName                           |
+----------------------------+-----------------------------------------------------------------------------------------------------------------------+----------------------------------+------------------------------------+
| 2019-08-08 15:46:32.760104 | --- start span trace ----                                                                                             |                                  | trace                              |
| 2019-08-08 15:46:32.760107 | --- start span session.Execute ----                                                                                   |                                  | session.Execute                    |
| 2019-08-08 15:46:32.760111 | execute: replace  into t2(a, b) select a, b from t                                                                    |                                  | session.Execute                    |
| 2019-08-08 15:46:32.760111 | --- start span session.getTxnFuture ----                                                                              |                                  | session.getTxnFuture               |
| 2019-08-08 15:46:32.760123 | --- start span session.ParseSQL ----                                                                                  |                                  | session.ParseSQL                   |
| 2019-08-08 15:46:32.760138 | --- start span executor.Compile ----                                                                                  |                                  | executor.Compile                   |
| 2019-08-08 15:46:32.760250 | --- start span session.runStmt ----                                                                                   |                                  | session.runStmt                    |
| 2019-08-08 15:46:32.760299 | --- start span executor.handleNoDelayExecutor ----                                                                    |                                  | executor.handleNoDelayExecutor     |
| 2019-08-08 15:46:32.760302 | --- start span *executor.ReplaceExec.Next ----                                                                        |                                  | *executor.ReplaceExec.Next         |
| 2019-08-08 15:46:32.760307 | --- start span *executor.TableReaderExecutor.Next ----                                                                |                                  | *executor.TableReaderExecutor.Next |
| 2019-08-08 15:46:32.760344 | table scan table: t, range: [[-inf,+inf]]                                                                             |                                  | *executor.TableReaderExecutor.Next |
| 2019-08-08 15:46:32.760364 | send Cop request to region 8 at 127.0.0.1:9191                                                                        |                                  | session.runStmt                    |
| 2019-08-08 15:46:32.760574 | receive *tikvpb.BatchCommandsResponse_Response_Coprocessor response with other 1 batched requests from 127.0.0.1:9191 |                                  | session.runStmt                    |
| 2019-08-08 15:46:32.760625 | --- start span *executor.TableReaderExecutor.Next ----                                                                |                                  | *executor.TableReaderExecutor.Next |
| 2019-08-08 15:46:32.760628 | table scan table: t, range: [[-inf,+inf]]                                                                             |                                  | *executor.TableReaderExecutor.Next |
| 2019-08-08 15:46:32.760672 | --- start span session.CommitTxn ----                                                                                 |                                  | session.CommitTxn                  |
| 2019-08-08 15:46:32.760673 | --- start span session.doCommitWitRetry ----                                                                          |                                  | session.doCommitWitRetry           |
| 2019-08-08 15:46:32.760695 | send Prewrite request to region 8 at 127.0.0.1:9191                                                                   | map[commitTs:410320998959415299] | session.doCommitWitRetry           |
| 2019-08-08 15:46:32.760828 | receive *tikvpb.BatchCommandsResponse_Response_Prewrite response with other 1 batched requests from 127.0.0.1:9191    | map[commitTs:410320998959415299] | session.doCommitWitRetry           |
| 2019-08-08 15:46:32.760833 | start get commit ts                                                                                                   | map[commitTs:410320998959415299] | session.doCommitWitRetry           |
| 2019-08-08 15:46:32.760897 | finish get commit ts                                                                                                  | map[commitTs:410320998959415299] | session.doCommitWitRetry           |
| 2019-08-08 15:46:32.760908 | send Commit request to region 8 at 127.0.0.1:9191                                                                     | map[commitTs:410320998959415299] | session.doCommitWitRetry           |
| 2019-08-08 15:46:32.763255 | receive *tikvpb.BatchCommandsResponse_Response_Commit response with other 1 batched requests from 127.0.0.1:9191      | map[commitTs:410320998959415299] | session.doCommitWitRetry           |
| 2019-08-08 15:46:32.763288 | execute done, modify row: 3                                                                                           |                                  | trace                              |
+----------------------------+-----------------------------------------------------------------------------------------------------------------------+----------------------------------+------------------------------------+
24 rows in set (0.00 sec)

```

we can see that happen and its property(what host? what region? what table?)

and what cause it be retry? how to retry? and retry which host? and so on...

we also can add more detail about: how optimizer choose two plan, apply one optimize rule cause plan tree change from a to b just [like this](https://gist.github.com/lysu/07249e14d0c06336e2b86ea58e7c1b01)  and only pay cost when we execute sql using trace key word

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - performance  just as normal trace

Related changes

 - n/a

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/11681)

### CODE REVIEW

**Comment by tiancaiamao** [comment]
LGTM

**Comment by cfzjywxk** [comment]
will this do some merge/aggregation work on some function calls many times, and show (total_count/total_time_usage)

File: executor/batch_checker.go [hunk 1]
```diff
@@ -14,6 +14,7 @@
 package executor
 
 import (
+	"context"
```
Comment (jackysp) [inline executor/batch_checker.go]: add a blank line below

File: util/logutil/log.go [hunk 2]
```diff
@@ -356,3 +358,27 @@ func WithKeyValue(ctx context.Context, key, value string) context.Context {
 	}
 	return context.WithValue(ctx, ctxLogKey, logger.With(zap.String(key, value)))
 }
+
+// TraceEventKey presents the TraceEventKey in span log.
+const TraceEventKey = "event"
+
+// Event records event in current tracing span.
+func Event(ctx context.Context, event string) {
```
Comment (cfzjywxk) [inline util/logutil/log.go]: do we need to abstract some "EventType"  params,  this maybe a "wait-start"、"wake-up incident" like trace point from different insert, select or DDL. Future we could make reports or generated views on these events based on different types?

File: store/tikv/2pc.go [hunk 3]
```diff
@@ -948,25 +951,27 @@ func (c *twoPhaseCommitter) checkSchemaValid() error {
...
 }
 
-func (c *twoPhaseCommitter) prewriteBinlog() chan error {
+func (c *twoPhaseCommitter) prewriteBinlog(ctx context.Context) chan error {
 	if !c.shouldWriteBinlog() {
 		return nil
 	}
 	ch := make(chan error, 1)
 	go func() {
+		logutil.Eventf(ctx, "start prewrite binlog")
```
Comment (cfzjywxk) [inline store/tikv/2pc.go]: abstract some PrewriteBegin event type?

**Comment by lysu** [comment]
> will this do some merge/aggregation work on some function calls many times, and show (total_count/total_time_usage)

good idea, but maybe that data are more suitable for `format=row` that aggregate at  span level, log just want to show event timeline

File: util/logutil/log.go (hunk 2 above)
Comment (lysu) [reply util/logutil/log.go]: good idea, how about do that after we design a full `EventType` category in the future?

**Pull request merged**